shell/platform/darwin/macos/framework/Source/FlutterMenuPlugin.mm

```diff
@@ -0,0 +1,420 @@
+                                                                 andChannel:_channel];
+    [_menuDelegates addObject:delegate];
+    [[NSApp.mainMenu itemAtIndex:index++] submenu].delegate = delegate;
+  }
+}
+
+- (NSMenuItem*)findDefaultMenu:(NSMenu*)menu ofType:(SEL)selector withTag:(int)tag {
+  auto enumerator = menu != nullptr ? menu.itemArray.objectEnumerator
+                                    : _platformProvidedItemArray.objectEnumerator;
+  for (NSMenuItem* item in enumerator) {
+    if ([item action] == selector && (tag == 0 || [item tag] == tag)) {
+      return item;
+    }
+    if ([[item submenu] numberOfItems] > 0) {
+      NSMenuItem* child_find = [self findDefaultMenu:[item submenu] ofType:selector withTag:tag];
+      if (child_find != nullptr) {
+        return child_find;
+      }
+    }
+  }
+  return nullptr;
+}
+
+- (NSMenuItem*)createPlatformProvidedMenu:(PlatformProvidedMenu)type {
+  NSMenuItem* found = nullptr;
+  auto found_type = kMacOSProvidedMenus.find(type);
+  if (found_type == kMacOSProvidedMenus.end()) {
+    return nullptr;
+  }
+  SEL selector_target = found_type->second;
+  // Since it doesn't have a definitive selector, the Services submenu is
+  // assumed to be the first item with a submenu action in the first menu item
+  // of the default menu set. We can't just get the title to check, since that
+  // is localized.
+  NSMenu* starting_menu = type == PlatformProvidedMenu::servicesSubmenu
+                              ? [_platformProvidedItemArray[0] submenu]
+                              : nullptr;
+  found = [self findDefaultMenu:starting_menu ofType:selector_target withTag:0];
+  if (found != nullptr) {
+    return [found copy];
+  }
+  return nullptr;
+}
+
+/**
+ * Constructs and returns an NSMenuItem corresponding to the item in
+ * |representation|, including recursively creating children if it has a
+ * submenu.
+ */
+- (NSMenuItem*)menuItemFromFlutterRepresentation:(NSDictionary*)representation {
+  if ([(NSNumber*)([representation valueForKey:kDividerKey]) intValue] == YES) {
+    return [NSMenuItem separatorItem];
+  } else {
+    NSString* appName = [NSRunningApplication currentApplication].localizedName;
+    NSString* title = representation[kLabelKey];
+    if (title != nil) {
+      title = [title stringByReplacingOccurrencesOfString:kAppName withString:appName];
+    }
+    NSNumber* boxedID = representation[kIdKey];
+    NSNumber* platformProvidedMenuId = representation[kPlatformProvidedMenuKey];
+    SEL action = (boxedID ? @selector(flutterMenuItemSelected:) : NULL);
+    NSString* keyEquivalent = @"";
+
+    if (platformProvidedMenuId != nil) {
+      return
+          [self createPlatformProvidedMenu:(PlatformProvidedMenu)platformProvidedMenuId.intValue];
+    } else {
+      NSNumber* triggerKeyId = representation[kShortcutTriggerKey];
+      if ([kMacOSSpecialKeys objectForKey:triggerKeyId] != nil) {
```

Comment:
Prefer `[]` to `objectForKey:`.

Also, consider making this a local variable rather than doing the lookup twice.

Reply:
That was just me learning ObjC containers. I did it below, just forgot to fix it everywhere when I learned that I didn't need to call the method. Deciding when ObjC wants its invocation syntax with `[foo bar]` vs regular C syntax was (is?) pretty confusing.